### PR TITLE
elixir: new packages

### PR DESCRIPTION
**Pull request by peterwillcn**
Signed-off-by: Xiaobo Tian peterwillcn@gmail.com

Maintainer: Xiaobo Tian peterwillcn@gmail.com
Compile tested: x86_64
Run tested: x86_64

Description:
Elixir is a dynamic, functional language designed for building scalable and maintainable applications.
Screen Shot 2020-04-21 at 17 07 08

### CODE REVIEW

**Comment by peterwillcn** [comment]
@BKPepe ping

File: lang/elixir/Makefile [hunk 1]
```diff
@@ -0,0 +1,69 @@
+#
+# Copyright (C) 2009-2020 OpenWrt.org
+#
+# This is free software, licensed under the GPL License v2.
+# See /LICENSE for more information.
+#
```
Comment (BKPepe) [inline lang/elixir/Makefile:6]: From https://github.com/openwrt/packages/blob/master/CONTRIBUTING.md

> Provide an up-to-date Copyright notice or none. Copyright should not be assigned to OpenWrt unless you are explicitly requested by or working under contract to OpenWrt. Assigning a Copyright to yourself or organization you represent is acceptable.

File: lang/elixir/Makefile [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+PKG_NAME:=elixir
+PKG_VERSION:=1.10.2
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-v$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://github.com/elixir-lang/elixir/archive/v$(PKG_VERSION)
+PKG_HASH:=5adffcf4389aa82fcfbc84324ebbfa095fc657a0e15b2b055fc05184f96b2d50
+
+PKG_MAINTAINER:=Xiaobo Tian <peterwillcn@gmail.com>
+PKG_LICENSE:=GPL-2.0
```
Comment (BKPepe) [inline lang/elixir/Makefile:19]: This is deprecated [SPDX License Identifier](https://spdx.org/licenses).

File: lang/elixir/Makefile [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+PKG_LICENSE_FILES:=LICENSE
+
+PKG_BUILD_PARALLEL:=1
+
+include $(INCLUDE_DIR)/package.mk
+
+PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_NAME)-$(PKG_VERSION)
```
Comment (BKPepe) [inline lang/elixir/Makefile:26]: This is already default. Remove it.

File: lang/elixir/Makefile [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  SUBMENU:=Elixir
+  SECTION:=lang
+  CATEGORY:=Languages
+  TITLE:=Elixir scripting language
+  URL:=http://elixir-lang.org
+endef
+
+define Package/elixir/Default/description
+	Elixir is a dynamic, functional language designed for building
+	scalable and maintainable applications
```
Comment (BKPepe) [inline lang/elixir/Makefile:38]: Two spaces as indentation.

File: lang/elixir/Makefile [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+include $(INCLUDE_DIR)/package.mk
+
+PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_NAME)-$(PKG_VERSION)
+
+define Package/elixir/Default
+  SUBMENU:=Elixir
+  SECTION:=lang
+  CATEGORY:=Languages
+  TITLE:=Elixir scripting language
+  URL:=http://elixir-lang.org
```
Comment (BKPepe) [inline lang/elixir/Makefile:33]:
```suggestion
  URL:=https://elixir-lang.org/
```

File: lang/elixir/Makefile [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+PKG_LICENSE:=GPL-2.0
+PKG_LICENSE_FILES:=LICENSE
+
+PKG_BUILD_PARALLEL:=1
+
+include $(INCLUDE_DIR)/package.mk
+
+PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_NAME)-$(PKG_VERSION)
+
+define Package/elixir/Default
```
Comment (BKPepe) [inline lang/elixir/Makefile:28]:
```suggestion
define Package/elixir
```

File: lang/elixir/Makefile [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  URL:=http://elixir-lang.org
+endef
+
+define Package/elixir/Default/description
+	Elixir is a dynamic, functional language designed for building
+	scalable and maintainable applications
+endef
+
+define Package/elixir
+$(call Package/elixir/Default)
```
Comment (BKPepe) [inline lang/elixir/Makefile:42]: Why? This is wrong. There is no other packages. Add your DEPENDS to section (row 28)

File: lang/elixir/Makefile [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  DEPENDS:=+erlang +erlang-compiler +erlang-runtime-tools +erlang-hipe
+endef
+
+define Package/elixir/description
+$(call Package/elixir/Default/description)
+ .
+ This package contains the runtime implementation and a minimal set of
+ modules.
+endef
```
Comment (BKPepe) [inline lang/elixir/Makefile:51]: Same as above.

File: lang/elixir/Makefile [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+define Build/Compile
+	$(MAKE) -C $(PKG_BUILD_DIR) \
+		PREFIX="$(PKG_INSTALL_DIR)/usr" \
+		install
+endef
```
Comment (BKPepe) [inline lang/elixir/Makefile:57]: Is this really needed?

Comment (peterwillcn) [reply lang/elixir/Makefile:57]: Yes, thereneed to compile install

File: lang/elixir/Makefile (hunk 8 above)
Comment (peterwillcn) [reply lang/elixir/Makefile:51]: Remove done.

File: lang/elixir/Makefile (hunk 7 above)
Comment (peterwillcn) [reply lang/elixir/Makefile:42]: Remove done.

File: lang/elixir/Makefile (hunk 6 above)
Comment (peterwillcn) [reply lang/elixir/Makefile:28]: fixed

File: lang/elixir/Makefile (hunk 5 above)
Comment (peterwillcn) [reply lang/elixir/Makefile:33]: Fixed

File: lang/elixir/Makefile (hunk 4 above)
Comment (peterwillcn) [reply lang/elixir/Makefile:38]: Fixed

File: lang/elixir/Makefile (hunk 3 above)
Comment (peterwillcn) [reply lang/elixir/Makefile:26]: Remove done

File: lang/elixir/Makefile (hunk 2 above)
Comment (peterwillcn) [reply lang/elixir/Makefile:19]: Updated

File: lang/elixir/Makefile (hunk 1 above)
Comment (peterwillcn) [reply lang/elixir/Makefile:6]: Updated

**Comment by peterwillcn** [comment]
@BKPepe ping

File: lang/elixir/Makefile (hunk 2 above)
Comment (neheb) [reply lang/elixir/Makefile]: Still no. It needs -or-later or -only.

**Comment by neheb** [comment]
I suggest using PKG_INSTALL instead of that Build/Compile section.

**Comment by peterwillcn** [comment]
@neheb I'm afraid I still haven't understood, what is the difference between PKG_INSTALL and Compile?

**Comment by neheb** [comment]
> @neheb I'm afraid I still haven't understood, what is the difference between PKG_INSTALL and Compile?

PKG_INSTALL is less verbose. Otherwise no difference.

File: lang/elixir/Makefile (hunk 9 above)
Comment (peterwillcn) [reply lang/elixir/Makefile:44]: The default compiled folder is adulterated with source code, so PKG_INSTALL_DIR needs to be distinguished to help reduce the package size.

**Comment by peterwillcn** [comment]
@neheb I tried PKG_INSTALL several times and failed. So keep using Build/Compile, or you can help with it. Thanks.

**Comment by neheb** [comment]
getting this warning during compilation:

```
warning: the VM is running with native name encoding of latin1 which may cause Elixir to malfunction as it expects utf8. Please ensure your locale is set to UTF-8 (which can be verified by running "locale" in your shell)
```

Not sure if a problem.

**Comment by peterwillcn** [comment]
@neheb ping

File: lang/elixir/Makefile [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+define Package/elixir/description
+  Elixir is a dynamic, functional language designed for building
+  scalable and maintainable applications
+endef
+
+define Build/Compile
+	$(MAKE) -C $(PKG_BUILD_DIR) \
+	PREFIX="$(PKG_INSTALL_DIR)/usr" \
+  install
+endef
```
Comment (dangowrt) [inline lang/elixir/Makefile:44]: As previously stated, please do not redefined `Build/Compile` in that way as the same thing can most likely be achieved by simply setting `PKG_INSTALL:=1` (eg. right under `PKG_BUILD_PARALLEL:=1`)

Comment (peterwillcn) [reply lang/elixir/Makefile]: Remove Build/Compile done

**Comment by dangowrt** [comment]
Please squash all commits into one single commit with `elixir: add new package` as title and your SoB, then force-push it into in this PR branch.

**Comment by dangowrt** [comment]
From watching the CI it seems like `make install` doesn't care about `DESTDIR`...

**Comment by peterwillcn** [comment]
> From watching the CI it seems like `make install` doesn't care about `DESTDIR`...

How to fix this?

**Comment by dangowrt** [comment]
> > From watching the CI it seems like `make install` doesn't care about `DESTDIR`...
> 
> How to fix this?

Usually setting PREFIX like you did before has some unwanted side-effects (as that patch might also be relevant in run-time).
Looking at the toplevel `Makefile` of Elixir, it does look like DESTDIR is handled correctly. Ie. in the OpenWrt package recipe you need to do
`$(CP) $(PKG_INSTALL_DIR)/usr/elixir/lib/* $(1)/usr/lib/elixir/lib/`
instead of
`$(CP) $(PKG_INSTALL_DIR)/elixir/lib/* $(1)/usr/lib/elixir/lib/`
or something like that.
Check the content of `build_dir/*/elixir*/ipkg-install/` to see what it actually installed there.

File: lang/elixir/Makefile [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+PKG_LICENSE:=Apache-2.0
+PKG_LICENSE_FILES:=LICENSE
+PKG_MAINTAINER:=Xiaobo Tian <peterwillcn@gmail.com>
+
+PKG_INSTALL:=1
+
+include $(INCLUDE_DIR)/package.mk
+
+define Package/elixir
+  SUBMENU:=Elixir
```
Comment (BKPepe) [inline lang/elixir/Makefile:27]: Is this really needed? In that submenu, there is going to be just one package.

Comment (peterwillcn) [reply lang/elixir/Makefile]: Remove SUBMENU done

**Comment by peterwillcn** [comment]
@dangowrt I don't know how to modify it. I need assistance.

**Comment by neheb** [comment]
@peterwillcn add PKG_INSTALL:=1 above PKG_BUILD_PARALLEL.

also, rebase as we now have some new CI.

**Comment by BKPepe** [comment]
Let's close it for now.

**Pull request closed without merging**